File: src/test/java/personal/alexgarland/helloworldmvc/controller/EmployeeActionControllerTest.java
package personal.alexgarland.helloworldmvc.controller;

import static java.util.Arrays.asList;
import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.springframework.web.servlet.ModelAndView;

import ch.qos.logback.classic.Level;
import personal.alexgarland.helloworldmvc.model.Employee;
import personal.alexgarland.helloworldmvc.service.IEmployeeRepository;
import personal.alexgarland.helloworldmvc.testutil.LogbackEventChecker;


public class EmployeeActionControllerTest {

	private static final Employee employee1 = new Employee(1, "Robert", "Dobalina");
	private static final Employee employee2 = new Employee(2, "Teren", "Jones");
	private static final List<Employee> employees = new ArrayList<Employee>(asList(employee1, employee2));
	
	private EmployeeActionController testController;
    
	@Mock
	private IEmployeeRepository employeeRepository;
	
    @Before
    public void setUp(){
    	
        MockitoAnnotations.initMocks(this);
		when(employeeRepository.getEmployeeList()).thenReturn(employees);
		when(employeeRepository.getEmployeeById(1)).thenReturn(employee1);
		
		testController = new EmployeeActionController();
		testController.setEmployeeRepository(employeeRepository);
		
		LogbackEventChecker.setup();
    
    }
    	
    @After
    public void teardown() {
    
    	LogbackEventChecker.teardown();
    
    }
	 
	@Test
	public void listsEmployees() {
		ModelAndView result = testController.getEmployeeList();
		
		assertEquals("employeeList", result.getViewName());
		
		Map<String, Object> model = result.getModel();
		assertEquals(1, model.size());
		assertEquals(employees, model.get("employeeList"));
		
		new LogbackEventChecker()
		.withExpectedEvent(Level.INFO, "Listing 2 employees...")
		.withExpectedEvent(Level.DEBUG, employee1.toString())
		.withExpectedEvent(Level.DEBUG, employee2.toString())		
		.verifyCalls();

	}
    
	@Test
	public void getsAddEmployeeForm() {
		ModelAndView result = testController.getAddEmployeeForm();
		Map<String, Object> model = result.getModel();
		
		assertEquals("employeeAdd", result.getViewName());
		assertEquals(1, model.size());
		assert(model.get("employeeEntity") instanceof Employee);
		
		new LogbackEventChecker()
		.withExpectedEvent(Level.INFO, "Setting up \"Add Employee\" form")
		.verifyCalls();
		
	}
    
	@Test
	public void addsEmployee() {
		ModelAndView result = testController.addEmployee(employee1);
		verify(employeeRepository, times(1)).addEmployee(employee1);
		assertEquals("redirect:employees", result.getViewName());
		assertEquals(0, result.getModel().size());
		
		new LogbackEventChecker()
		.withExpectedEvent(Level.INFO, "Adding employee:")
		.withExpectedEvent(Level.INFO, employee1.toString())
		.verifyCalls();
		
	}
	
	@Test
	public void deletesEmployee() {
		testController.deleteEmployeeById(3);
		verify(employeeRepository, times(1)).deleteEmployeeById(3);
		
		new LogbackEventChecker()
		.withExpectedEvent(Level.WARN, "Deleting employee with ID 3")
		.verifyCalls();
		
	}

	@Test
	public void getsUpdateEmployeeForm() {
		ModelAndView result = testController.getUpdateEmployeeForm(1);
		Map<String, Object> model = result.getModel();
		
		assertEquals("employeeUpdate", result.getViewName());
		assertEquals(1, model.size());
		assertEquals(employee1, model.get("employeeEntity"));
		
		new LogbackEventChecker()
		.withExpectedEvent(Level.INFO, "Setting up Employee Update form")
		.verifyCalls();
		
	}

	@Test
	public void updatesEmployee() {
		Employee employee1NewDetails = new Employee(1, "Robert", "Dobalina", "Bob");
		testController.updateEmployee(employee1NewDetails);
		verify(employeeRepository, times(1)).updateEmployee(employee1NewDetails);
		
		new LogbackEventChecker()
		.withExpectedEvent(Level.INFO, "Updating employee with details:")
		.withExpectedEvent(Level.INFO, employee1NewDetails.toString())
		.verifyCalls();
		
	}
	
}

File: src/test/kotlin/personal/alexgarland/helloworldmvc/controller/EmployeeActionControllerTest.kt
package personal.alexgarland.helloworldmvc.controller

import ch.qos.logback.classic.Level
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.MockitoAnnotations
import personal.alexgarland.helloworldmvc.model.Employee
import personal.alexgarland.helloworldmvc.service.repository.IEmployeeRepository
import personal.alexgarland.helloworldmvc.testutil.LogbackEventChecker
import java.util.*
import java.util.Arrays.asList

class EmployeeActionControllerTest {

    private var testController: EmployeeActionController? = null

    @Mock
    private val employeeRepository: IEmployeeRepository? = null

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        `when`(employeeRepository!!.employeeList).thenReturn(employees)
        `when`(employeeRepository.getEmployeeById(1)).thenReturn(employee1)

        testController = EmployeeActionController(employeeRepository)

        LogbackEventChecker.setup()
    }

    @After
    fun teardown() = LogbackEventChecker.teardown()

    @Test
    fun listsEmployees() {
        val result = testController!!.employeeList

        assertEquals("employeeList", result.viewName)

        val model = result.model
        assertEquals(1, model.size.toLong())
        assertEquals(employees, model["employeeList"])

        LogbackEventChecker()
                .withExpectedEvent(Level.INFO, "Listing 2 employees...")
                .withExpectedEvent(Level.DEBUG, employee1.toString())
                .withExpectedEvent(Level.DEBUG, employee2.toString())
                .verifyCalls()

    }

    @Test
    fun getsAddEmployeeForm() {
        val result = testController!!.addEmployeeForm
        val model = result.model

        assertEquals("employeeAdd", result.viewName)
        assertEquals(1, model.size.toLong())
        assert(model["employeeEntity"] is Employee)

        LogbackEventChecker()
                .withExpectedEvent(Level.INFO, "Setting up \"Add Employee\" form")
                .verifyCalls()

    }

    @Test
    fun addsEmployee() {
        val result = testController!!.addEmployee(employee1)
        verify<IEmployeeRepository>(employeeRepository, times(1)).addEmployee(employee1)
        assertEquals("redirect:employees", result.viewName)
        assertEquals(0, result.model.size.toLong())

        LogbackEventChecker()
                .withExpectedEvent(Level.INFO, "Adding employee:")
                .withExpectedEvent(Level.INFO, employee1.toString())
                .verifyCalls()

    }

    @Test
    fun deletesEmployee() {

        testController!!.deleteEmployeeById(3)

        verify<IEmployeeRepository>(employeeRepository, times(1))
                .deleteEmployeeById(3)

        LogbackEventChecker()
                .withExpectedEvent(Level.WARN, "Deleting employee with ID 3")
                .verifyCalls()

    }

    @Test
    fun getsUpdateEmployeeForm() {
        val result = testController!!.getUpdateEmployeeForm(1)
        val model = result.model

        assertEquals("employeeUpdate", result.viewName)
        assertEquals(1, model.size.toLong())
        assertEquals(employee1, model["employeeEntity"])

        LogbackEventChecker()
                .withExpectedEvent(Level.INFO, "Setting up Employee Update form")
                .verifyCalls()

    }

    @Test
    fun updatesEmployee() {

        val employee1NewDetails = Employee(1, "Robert", "Dobalina", "Bob")
        testController!!.updateEmployee(employee1NewDetails)

        verify<IEmployeeRepository>(employeeRepository, times(1))
                .updateEmployee(employee1NewDetails)

        LogbackEventChecker()
                .withExpectedEvent(Level.INFO, "Updating employee with details:")
                .withExpectedEvent(Level.INFO, employee1NewDetails.toString())
                .verifyCalls()

    }

    companion object {
        private val employee1 = Employee(1, "Robert", "Dobalina")
        private val employee2 = Employee(2, "Teren", "Jones")
        private val employees = ArrayList(asList(employee1, employee2))
    }

}
